drivers: dac_stm32: add support for internally connected channels

xodus7 opened the pull request:
Adds support for configuring the STM32 DAC modes for internal connections. This is only enabled for SOCs that support it. As
part of this change, we make pinctrl configuration in the device tree optional.

xodus7 committed:
drivers: dac_stm32: make pinctrl config in the device tree optional

Now that we support internally connected channels we should make the
pinctrl configuration optional.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

martinjaeger reviewed (changes requested):
Thanks @xodus7 for your PR.

The changes look good to me. However, I think we should also go through all other drivers and make `dac_channel_setup` return `-ENOTSUP` if the HW doesn't support internal connections or `-ENOSYS` if that feature is not yet implemented if `.internal` is set to `true`. (I realize we should have done this for the `.buffered` flag as well)

And may I ask you to move the change to `include/zephyr/drivers/dac.h` into a dedicated commit, as it affects the entire DAC API and not only the STM32 driver?

xodus7 committed:
drivers: dac: add flag to channel config to support internal connections

Allows user to connect DAC output to internal peripherals for HW and
drivers that support it.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

xodus7 committed:
drivers: dac_stm32: add support for internally connected channels

Adds support for configuring the STM32 DAC channel mode for internal
connections. This is only enabled for SOCs that support it.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

xodus7 committed:
drivers: dac: add checks for internal channel config

Return -ENOTSUP if these flags are specified but not supported.

Signed-off-by: Corey Wharton <xodus7@cwharton.com>

xodus7 commented:
> The changes look good to me. However, I think we should also go through all other drivers and make `dac_channel_setup` return `-ENOTSUP` if the HW doesn't support internal connections or `-ENOSYS` if that feature is not yet implemented if `.internal` is set to `true`. (I realize we should have done this for the `.buffered` flag as well)
>

I pushed an update that adds checks for both in all drivers. I'm not 100% sure about the dac_mcux_gau.c driver which might need closer review from someone with more familiarity with that HW, but I tried to preserve existing functionality while supporting the new internal flag.
 
> And may I ask you to move the change to `include/zephyr/drivers/dac.h` into a dedicated commit, as it affects the entire DAC API and not only the STM32 driver?

Done.

martinjaeger commented:
> > The changes look good to me. However, I think we should also go through all other drivers and make `dac_channel_setup` return `-ENOTSUP` if the HW doesn't support internal connections or `-ENOSYS` if that feature is not yet implemented if `.internal` is set to `true`. (I realize we should have done this for the `.buffered` flag as well)
> 
> I pushed an update that adds checks for both in all drivers. I'm not 100% sure about the dac_mcux_gau.c driver which might need closer review from someone with more familiarity with that HW, but I tried to preserve existing functionality while supporting the new internal flag.

@decsny Could you please take a look?

martinjaeger reviewed (changes requested):
I have double-checked the hardware capabilities of some DACs and it's not easy as it seems. Most external DACs seem to have a built-in output buffer which cannot be disabled, whereas internal DACs sometimes have buffers which must be used for connections to external pins.

Probably it's easier if you just add the check for the newly added `internal` flag and leave out the `buffered` flag for now.

xodus7 commented:
> Probably it's easier if you just add the check for the newly added `internal` flag and leave out the `buffered` flag for now.

Ok, fair, I reverted most of the checks for buffered flag. Thanks for looking closely at this!

decsny commented:
it's worth noting that I think NXP has DACs like this on chips where the output can be used as an internal reference, I think some of which are in tree already

also, I noticed this is a similar thing to how someone proposed something to the counter api and dma api to enable internal trigger connections, maybe some generic solution will be desired someday instead of every subsystem having it's own api to route and enable internal chip connections

danieldegrasse commented:
> it's worth noting that I think NXP has DACs like this on chips where the output can be used as an internal reference, I think some of which are in tree already

> also, I noticed this is a similar thing to how someone proposed something to the counter api and dma api to enable internal trigger connections, maybe some generic solution will be desired someday instead of every subsystem having it's own api to route and enable internal chip connections

I kind of think making this value driver dependent makes more sense- something like how the ADC API works: https://github.com/zephyrproject-rtos/zephyr/blob/86af9c4b0766842aa97b5b9bead12c71ca5d69ed/include/zephyr/drivers/adc.h#L141 This would better handle cases where the DAC might have multiple internal sources it can route to.

@xodus7 thoughs on this approach?

xodus7 commented:
> This would better handle cases where the DAC might have multiple internal sources it can route to.
> @xodus7 thoughs on this approach?

Yeah that makes sense to me and is not a big change. Let me make a modification.

martinjaeger commented:
> I kind of think making this value driver dependent makes more sense- something like how the ADC API works:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/86af9c4b0766842aa97b5b9bead12c71ca5d69ed/include/zephyr/drivers/adc.h#L141
> 
> This would better handle cases where the DAC might have multiple internal sources it can route to.
> @xodus7 thoughs on this approach?

@danieldegrasse not sure I fully understand your suggestion. I'm not very familiar with the NXP MCUs, but for STM32 we would configure the DAC for internal output (nothing else to be configured) and afterwards you'd configure other peripherals (e.g. comparators) to use the DAC as their source. But that would belong to the driver of the other peripheral, IMO. So I'm not sure what sort of information we would potentially store in the `.internal` flag if it's not a `bool`. Do you have an example?

danieldegrasse commented:
> @danieldegrasse not sure I fully understand your suggestion. I'm not very familiar with the NXP MCUs, but for STM32 we would configure the DAC for internal output (nothing else to be configured) and afterwards you'd configure other peripherals (e.g. comparators) to use the DAC as their source. But that would belong to the driver of the other peripheral, IMO. So I'm not sure what sort of information we would potentially store in the .internal flag if it's not a bool. Do you have an example?

I don't unfortunately- none of the NXP DACs seem to have the feature I'm thinking this would enable, so we are likely fine with a `bool`. I'm envisioning a case where a DAC can be internally routed as an input to several peripherals, and that selection is made by a register on the DAC. That being said, I'm not sure *which* peripherals beyond a comparator would really need a DAC. It seems like making this flag more than a `bool` might be overkill- especially since the DAC API is unstable, so this can change in the future.

xodus7 commented:
> I don't unfortunately- none of the NXP DACs seem to have the feature I'm thinking this would enable, so we are likely fine with a `bool`. I'm envisioning a case where a DAC can be internally routed as an input to several peripherals, and that selection is made by a register on the DAC. That being said, I'm not sure _which_ peripherals beyond a comparator would really need a DAC. It seems like making this flag more than a `bool` might be overkill- especially since the DAC API is unstable, so this can change in the future.

It seems totally reasonable to me to spend a few bits to future proof things but I'm also not aware of any DAC peripheral that operates this way. If the DAC API is considered unstable I guess it doesn't really matter either way. Please advise.

danieldegrasse commented:
> It seems totally reasonable to me to spend a few bits to future proof things but I'm also not aware of any DAC peripheral that operates this way. If the DAC API is considered unstable I guess it doesn't really matter either way. Please advise.

I'm not the authority here, but personally I'm convinced we are ok to keep it as a `bool` given that I can't find a DAC that works the way I described.
